File: internal/parser/tables.go
package parser

import (
	"strings"
)

func tables(input string) (node, error) {
	lines := strings.Split(strings.TrimSpace(input), "\n")
	headRow := lines[0]
	alignRow := lines[1]
	tbody := lines[2:]

	thead := strings.Split(headRow[1:len(headRow)-1], "|")
	for i := range thead {
		thead[i] = strings.TrimSpace(thead[i])
	}

	alignments := strings.Split(alignRow[1:len(alignRow)-1], "|")
	for i := range alignments {
		col := strings.TrimSpace(alignments[i])
		switch {
		case col[0] == ':' && col[len(col)-1] == ':':
			alignments[i] = "center"
		case col[len(col)-1] == ':':
			alignments[i] = "right"
		case col[0] == ':':
			alignments[i] = "left"
		default:
			alignments[i] = ""
		}
	}

	theadNodes := []node{}
	for i, th := range thead {
		children, err := inlineFormatting(th)
		if err != nil {
			return nil, err
		}
		theadNodes = append(theadNodes, element{"th", map[string]string{"align": alignments[i]}, children})
	}

	tbodyNodes := []node{}
	for _, line := range tbody {
		trNodes := []node{}
		for i, col := range strings.Split(line[1:len(line)-1], "|") {
			children, err := inlineFormatting(strings.TrimSpace(col))
			if err != nil {
				return nil, err
			}
			trNodes = append(trNodes, element{"td", map[string]string{"align": alignments[i]}, children})
		}
		tbodyNodes = append(tbodyNodes, element{TagName: "tr", Children: trNodes})
	}

	return element{TagName: "table", Children: []node{
		element{TagName: "thead", Children: []node{
			element{TagName: "tr", Children: theadNodes},
		}},
		element{TagName: "tbody", Children: tbodyNodes},
	}}, nil
}

File: crates/znak/src/parser/tables.rs
use crate::parser::{
    inline_formatting::inline_formatting,
    types::{Node, element},
};

pub(crate) fn tables(input: &str) -> Node {
    let lines = input.trim().lines().collect::<Vec<&str>>();
    let head_row = lines[0];
    let align_row = lines[1];
    let tbody = lines[2..].to_vec();

    let thead = head_row[1..head_row.len() - 1]
        .split("|")
        .map(|item| item.trim());

    let alignments = align_row[1..align_row.len() - 1]
        .split("|")
        .map(|col| {
            let col = col.trim();
            match (col.starts_with(":"), col.ends_with(":")) {
                (true, true) => "center",
                (false, true) => "right",
                (true, false) => "left",
                _ => "",
            }
        })
        .collect::<Vec<&str>>();

    let mut thead_nodes = vec![];
    for (i, th) in thead.enumerate() {
        let children = inline_formatting(th.to_string());
        thead_nodes.push(element!("th", [align = alignments[i]], children));
    }

    let mut tbody_nodes = vec![];
    for line in tbody {
        let mut tr_nodes = vec![];
        for (i, col) in line[1..line.len() - 1].split("|").enumerate() {
            let children = inline_formatting(col.trim().to_string());
            tr_nodes.push(element!("td", [align = alignments[i]], children));
        }
        tbody_nodes.push(element!("tr", tr_nodes));
    }

    element!(
        "table",
        vec![
            element!("thead", vec![element!("tr", thead_nodes)]),
            element!("tbody", tbody_nodes),
        ]
    )
}
